### PR TITLE
fixed:generator method of class autocomplete causes star(*) to be add…

### LINKED ISSUE #43469: generator method of class autocomplete causes star(*) to be added to next line
vscode

- VSCode Version: 1.20.0
- OS Version:  macos 10.13.3

Steps to Reproduce:

1.  write one class 
2.  add generator function

image

Does this issue occur when all extensions are disabled?: Yes

### PR DESCRIPTION
**Pull request by limerickgds**
i add a onEnterRule rule that will only execute if the text above the this line matches /** or *.
but i don't think this is a good solution.
Fixes #43469
cc: @joaomoreno

### CODE REVIEW

**Comment by limerickgds** [comment]
@misolori do you need anything from me to merge this PR?

**Comment by misolori** [comment]
@limerickgds I'll let @aeschli handle this PR

**Comment by mjbvz** [comment]
@alexandrudima / @rebornix Can you please take a look at this

**Comment by limerickgds** [comment]
@mjbvz @misolori @aeschli  How long do I have to wait?

**Comment by aeschli** [comment]
@alexandrudima and @rebornix have the final word here, but looking at the previous line looks fishy to me. For other language configurations we can already express where the rule applies:
```
	"autoClosingPairs": [
		{ "open": "/*", "close": "*/", "notIn": ["string"] },
		{ "open": "\"", "close": "\"", "notIn": ["string", "comment"] },
	]
```
See https://github.com/Microsoft/vscode/blob/7c03ce156505cca6c60bd5afd9b2f33f25ec64ad/src/vs/editor/common/modes/languageConfiguration.ts#L177.

Ideally, we can add the same for `OnEnterRule`

**Comment by alexandrudima** [comment]
I personally kind of like it that the implemention of `onEnterRules` is kept tokenization agnostic and I think the new rule (`oneLineAboveText`) is in the same spirit with my original implemention (i.e. similar to `beforeText` and `afterText`).

@limerickgds Thank you !

File: src/vs/vscode.d.ts [hunk 1]
```diff
@@ -3738,6 +3738,10 @@ declare module 'vscode' {
 		 * This rule will only execute if the text after the cursor matches this regular expression.
 		 */
 		afterText?: RegExp;
+		/**
+		 * This rule will only execute if the text above the this line matches this regular expression.
+		 */
+		oneLineAboveText?: RegExp;
```
Comment (alexandrudima) [inline src/vs/vscode.d.ts]: fyi @jrieken this PR adds a new property to `OnEnterRule` in `vscode.d.ts`.

Comment (jrieken) [reply src/vs/vscode.d.ts]: @alexandrudima please create a tracking item so we know this is an API change

**Review by jrieken — COMMENTED** [review]
.

**Pull request merged**